### ISSUE TITLE
ENH: spatial.distance: array API standard support tracker

### ISSUE THREAD

**Issue by amacati** [scipy/scipy#25734]
## Summary
Towards #18867. Since we are nearing 2.0 and there was still no effort for the conversion of the `spatial` module I thought I might give this one a shot.

## Architecture
There are two classes of functions here, the single-vector ones and the batched ones in `pdist` and `cdist`. The batched versions each have dedicated C++ bindings that will be hard to beat with the array API. 

To avoid regressions, I propose to follow a similar architecture as in the `spatial.transform` rewrite from #22500. We would have a backend that reimplements in native xp, and keep the C++ functions around. To avoid implementing dedicated batched functions, the xp backend should be generic over axes from the start. The single-vector distance functions will remain for compatibility, and keep the same checks as now, but then call the generic backend. 

`pdist` and `cdist` can call the same backend with batched inputs. The beauty of having a batch-native backend is that we can first convert the single-case functions one-by-one, review them, and, once we move on to `pdist` and `cdist`, everything should just fall into place. In addition, should we ever feel like exposing the fully batched versions, we can dispatch to the generic backend for numpy arrays of dim>2, and other frameworks should work by default.

## Converted
- [ ] Basic machinery for sanitizing and input checking: #25753 
- [ ] `minkowski`: #25753 
- [ ] `euclidean`
- [ ] `sqeuclidean`
- [ ] `correlation`
- [ ] `cosine`
- [ ] `hamming`
- [ ] `jaccard`
- [ ] `seuclidean`
- [ ] `cityblock`
- [ ] `mahalanobis`
- [ ] `chebyshev`
- [ ] `braycurtis`
- [ ] `canberra`
- [ ] `jensenshannon`
- [ ] `yule`
- [ ] `dice`
- [ ] `rogerstanimoto`
- [ ] `russellrao`
- [ ] `sokalsneath`
- [ ] `pdist`
- [ ] `squareform`
- [ ] `is_valid_dm`
- [ ] `is_valid_y`
- [ ] `num_obs_dm`
- [ ] `num_obs_y`
- [ ] `cdist`

## Scope decisions
- [ ] `directed_hausdorff`: Hard to convert since it requires randomness and early stopping. Might have to fall back to the exhaustive implementation

## Discussion
@lucascolley @j-bowhay Not sure whom else to ping for this discussion. Are you generally okay with this plan?

**Comment by lucascolley** [comment]
> To avoid regressions, I propose to follow a similar architecture as in the `spatial.transform` rewrite from [#22500](https://github.com/scipy/scipy/issues/22500). We would have a backend that reimplements in native xp, and keep the C++ functions around. To avoid implementing dedicated batched functions, the xp backend should be generic over axes from the start. The single-vector distance functions will remain for compatibility, and keep the same checks as now, but then call the generic backend.

> @lucascolley @j-bowhay Not sure whom else to ping for this discussion. Are you generally okay with this plan?

Plan sounds good overall to me. The only thing I would check would be whether https://github.com/scipy/scipy/tree/main/scipy/spatial/src would potentially fall under the scope of https://github.com/scipy/xsf/issues/220, cc @steppi . However, even if it would, it's probably worth executing your plan anyway as long as there aren't any big blockers and you have the energy to see it through — there are definitely benefits to having everything in terms of the array API that we miss out on going the gufunc route.

**Comment by lucascolley** [comment]
ping @rgommers @ev-br as well for vis

**Comment by amacati** [comment]
You mean if we can avoid dedicated C++ backends for numpy in favor of moving them to `xsf`? That might make sense, though @steppi can probably say a lot more about this. I would still write the generic backend though. `xsf` is opaque to other frameworks, hence auto-diff does not work. However, distance functions are especially likely to be used in combination with auto-diff, which we would get for free from an xp backend.

**Comment by lucascolley** [comment]
yeah, you said what I was trying to get across in a few words

**Comment by rgommers** [comment]
> The batched versions each have dedicated C++ bindings that will be hard to beat with the array API.

Performance is good but not great, see gh-19454 for what's possible with (a lot) more effort.

They're also in CuPy (with the optional cuVS dependency), and I think we haven't compared that nor are dispatching to it: https://docs.cupy.dev/en/stable/reference/scipy_spatial_distance.html. Would be nice to see that comparison.

For a native implementation, these functions are fairly poor fits I'd think? With eager execution it should be really slow. And I'm not sure how good of a job JITs will do. It'd be useful to try this for a single function, and see what `jax.jit` makes of it?

**Comment by ev-br** [comment]
+1 for all of trying a "generic backend" architecture, for comparing to the CuPy delegation target, and for checking jit vs eager. I too expect that eager will be pretty slow but jit might be competitive, similar to what we saw for RBF (which is, in a nutshell, a pairwise computation with a couple of additions) in https://labs.quansight.org/blog/array-api-aot-jit

**Comment by amacati** [comment]
> They're also in CuPy (with the optional cuVS dependency), and I think we haven't compared that nor are dispatching to it: https://docs.cupy.dev/en/stable/reference/scipy_spatial_distance.html. Would be nice to see that comparison.

Oh nice, I didn't know these existed. Should be easy to compare against and include in the dispatch, possibly dynamically based on the installed deps. Will run my benchmarks as usual when I have something to share.

**Comment by tylerjereddy** [comment]
> For a native implementation, these functions are fairly poor fits I'd think?

I think this is the closest comment to my current view/concern. I've put more detailed comments in the first prototype PR (gh-25753). I'm currently a bit negative, but I think the main thing I want to see is just the practical utility for the `xp` implementations vs. `pdist`/`cdist`. Is there a good broadcasting example where GPU with the individual distance functions really is better/more practical than `cdist` + slicing or whatever? I just haven't see the use case/examples yet maybe.

**Comment by steppi** [comment]
> `xsf` is opaque to other frameworks, hence auto-diff does not work. However, distance functions are especially likely to be used in combination with auto-diff, which we would get for free from an xp backend.

 `xsf`'s potential applicability here is a ways off, so shouldn't enter into consideration yet. I just want to point out that when `HiJAX`, https://docs.jax.dev/en/latest/ffi.html#supporting-transformations-like-vmap-and-grad, becomes part of JAX's stable API, we actually will be able to support autograd through functions backed by `xsf` kernels. I'm not sure when that may be, and the [jax cupy bridge](https://github.com/steppi/jax_cupy_bridge/tree/main) I've been working on that allows jitting around xsf-backed functions isn't even production ready yet, but autograd is on the roadmap. It won't come for free, but I don't think the effort would be too great either.

**Comment by tylerjereddy** [comment]
It might be good to see some brief discussion of space complexity too. I touched on it over at https://github.com/scipy/scipy/pull/25753#issuecomment-5138583252, but doesn't the "pdist as xp broadcasting" end up needing to discard a subset of the calculated values? Maybe the upper or lower triangle? Or maybe I'm just missing something...

**Comment by amacati** [comment]
> It might be good to see some brief discussion of space complexity too. I touched on it over at https://github.com/scipy/scipy/pull/25753#issuecomment-5138583252, but doesn't the "pdist as xp broadcasting" end up needing to discard a subset of the calculated values? Maybe the upper or lower triangle? Or maybe I'm just missing something...

I started working on a thorough evaluation with JAX on `pdist` and `cdist`. I can add memory consumption to it as well. Will share it once it is done, but it might take a bit longer because things are busy at the moment.

**Comment by amacati** [comment]
But no, you are not missing something. `cdist` is not doing unnecessary computations, so there things are fine. `pdist` is trickier. It either allocates more memory than necessary by building the full explicit matrices of the combinations, or calculates more and then drops the lower triangle. However, some lazy backends are smart enough to optimize this away. I still need to confirm this more definitively, but JAX seems to avoid the materialization of the triangle select on the GPU with `pmap`.
